File: src-lang/xt/db/system/impl_memory.clj
(ns xt.db.system.impl-memory
  (:require [hara.lang :as l]
            [xt.lang.common-protocol :as proto :refer [defimpl.xt]]))

(l/script :xtalk
  {:require [[xt.db.system.impl-common :as impl-common]
             [xt.db.system.memory-util :as util]
             [xt.db.system.memory-graph :as graph]
             [xt.db.text.base-flatten :as f]
             [xt.lang.common-protocol :as proto]
             [xt.lang.spec-base :as xt]
             [xt.lang.spec-promise :as promise]]})

(defn.xt pull
  "fetches tree ir data from the memory impl"
  {:added "4.1"}
  [impl tree]
  (var #{rows schema opts} impl)
  (return (graph/pull rows
                      schema
                      tree
                      opts)))

(defn.xt pull-async
  "fetches tree ir data with async semantics"
  {:added "4.1"}
  [impl tree]
  (return
   (promise/x:promise-run
    (-/pull impl tree))))

(defn.xt record-add
  "adds records directly to a single table in the memory impl"
  {:added "4.1"}
  [impl table-name records]
  (var #{rows schema opts} impl)
  (return
   (util/add-bulk rows schema {table-name records})))


(defn.xt record-delete
  "deletes ids directly from a single table in the memory impl"
  {:added "4.1"}
  [impl table-name ids]
  (var #{rows schema opts} impl)
  (return (util/remove-bulk rows
                            schema
                            table-name
                            ids)))


;;
;; PROCESS
;;

(defn.xt process-add-event
  [impl data]
  (var #{rows schema} impl)
  (return
   (util/add-bulk rows schema data)))

(defn.xt process-remove-event
  [impl data]
  (var #{rows schema lookup} impl)
  (var ordered (f/flatten-bulk-ids schema lookup data))
  (xt/for:array [entry ordered]
    (var [table-name ids] entry)
    (util/remove-bulk rows schema table-name ids))
  (return (xt/x:arr-map ordered xt/x:first)))

(defn.xt clear-db
  "clears all rows from the memory impl"
  {:added "4.1"}
  [impl]
  (var #{rows} impl)
  (xt/for:array [table-key (xt/x:obj-keys rows)]
    (xt/x:del-key rows table-key))
  (return nil))

(defn.xt rpc-call-async
  "memory impl does not support remote rpc calls"
  {:added "4.1"}
  [_impl _rpc-spec _args]
  (xt/x:err "db.impl.memory does not support rpc_call_async"))

;;
;; IMPL
;;

(defimpl.xt ImplMemory
  [rows schema lookup listeners]

  impl-common/ISourceLocal
  {impl-common/clear-db             -/clear-db
   impl-common/pull                 -/pull
   impl-common/record-add           -/record-add
   impl-common/record-delete        -/record-delete
   impl-common/process-add-event    -/process-add-event
   impl-common/process-remove-event -/process-remove-event}

  impl-common/ISourceRemote
  {impl-common/pull-async      -/pull-async
   impl-common/rpc-call-async  -/rpc-call-async}

  impl-common/ISourceListener
  {impl-common/add-db-listener     impl-common/add-db-listener-default
   impl-common/remove-db-listener  impl-common/remove-db-listener-default
   impl-common/get-db-listener     impl-common/get-db-listener-default})

(defn.xt impl-memory
  [schema lookup]
  (return
   (-/ImplMemory {} schema lookup {})))

File: packages-gen/dart/libs/xt-db/lib/system/impl-memory.dart
import 'package:xtalk_lang/common-protocol.dart' as proto;
import 'package:xtalk_db/system/memory-graph.dart' as graph;
import 'package:xtalk_db/system/memory-util.dart' as util;
import 'package:xtalk_db/system/impl-common.dart' as impl_common;
import 'package:xtalk_db/text/base-flatten.dart' as f;
import 'dart:async';






ImplMemory(rows, schema, lookup, listeners, metadata) {
  (__globals__["xt.lang.common_protocol/IMPLEMENTATIONS"] ??= <dynamic, dynamic>{})["xt.db.system.impl_memory/ImplMemory"] = true;
  proto.register_protocol_impl(impl_common.ISourceLocal["on"],"xt.db.system.impl_memory/ImplMemory",<dynamic, dynamic>{
    "clear_db":clear_db,
    "pull":pull,
    "record_add":record_add,
    "record_delete":record_delete,
    "process_add_event":process_add_event,
    "process_remove_event":process_remove_event
  });
  proto.register_protocol_impl(
    impl_common.ISourceRemote["on"],
    "xt.db.system.impl_memory/ImplMemory",
    <dynamic, dynamic>{"pull_async":pull_async,"rpc_call_async":rpc_call_async}
  );
  proto.register_protocol_impl(impl_common.ISourceListener["on"],"xt.db.system.impl_memory/ImplMemory",<dynamic, dynamic>{
    "add_db_listener":impl_common.add_db_listener_default,
    "remove_db_listener":impl_common.remove_db_listener_default,
    "get_db_listener":impl_common.get_db_listener_default
  });
  return <dynamic, dynamic>{
    "::":"xt.db.system.impl_memory/ImplMemory",
    "::/protocols":<dynamic>[
        impl_common.ISourceLocal["on"],
        impl_common.ISourceRemote["on"],
        impl_common.ISourceListener["on"]
      ],
    "::/protocol-impls":<dynamic, dynamic>{
        impl_common.ISourceLocal["on"]:<dynamic, dynamic>{
          "clear_db":clear_db,
          "pull":pull,
          "record_add":record_add,
          "record_delete":record_delete,
          "process_add_event":process_add_event,
          "process_remove_event":process_remove_event
        },
        impl_common.ISourceRemote["on"]:<dynamic, dynamic>{"pull_async":pull_async,"rpc_call_async":rpc_call_async},
        impl_common.ISourceListener["on"]:<dynamic, dynamic>{
          "add_db_listener":impl_common.add_db_listener_default,
          "remove_db_listener":impl_common.remove_db_listener_default,
          "get_db_listener":impl_common.get_db_listener_default
        }
      },
    "rows":rows,
    "schema":schema,
    "lookup":lookup,
    "listeners":listeners,
    "metadata":metadata
  };
}

pull(impl, tree) {
  var opts = impl["opts"];
  var rows = impl["rows"];
  var schema = impl["schema"];
  return graph.pull(rows,schema,tree,opts);
}

pull_async(impl, tree) {
  return Future.sync(() {
    return pull(impl,tree);
  });
}

record_add(impl, table_name, records) {
  var opts = impl["opts"];
  var rows = impl["rows"];
  var schema = impl["schema"];
  return util.add_bulk(rows,schema,<dynamic, dynamic>{table_name:records});
}

record_delete(impl, table_name, ids) {
  var opts = impl["opts"];
  var rows = impl["rows"];
  var schema = impl["schema"];
  return util.remove_bulk(rows,schema,table_name,ids);
}

process_add_event(impl, data) {
  var rows = impl["rows"];
  var schema = impl["schema"];
  return util.add_bulk(rows,schema,data);
}

process_remove_event(impl, data) {
  var lookup = impl["lookup"];
  var rows = impl["rows"];
  var schema = impl["schema"];
  var ordered = f.flatten_bulk_ids(schema,lookup,data);
  var arr_52126 = ordered;
  for(var i52127 = 0; i52127 < arr_52126.length; ++i52127){
    var entry = arr_52126[i52127];
    var value_52148 = entry;
    var table_name = value_52148[0];
    var ids = value_52148[1];
    util.remove_bulk(rows,schema,table_name,ids);
  };
  return xt.lang.common_data.arr_map(ordered,(arr) {
    return arr[0];
  });
}

clear_db(impl) {
  var rows = impl["rows"];
  var arr_52149 = List<dynamic>.from(( rows ).keys);
  for(var i52150 = 0; i52150 < arr_52149.length; ++i52150){
    var table_key = arr_52149[i52150];
    rows.remove(table_key);
  };
  return null;
}

rpc_call_async(_impl, _rpc_spec, _args) {
  throw "db.impl.memory does not support rpc_call_async";
}

impl_memory(schema, lookup) {
  return ImplMemory(
    <dynamic, dynamic>{},
    schema,
    lookup,
    <dynamic, dynamic>{},
    <dynamic, dynamic>{}
  );
}